Full CICD deploy with working yarn deploy. Solutions Builder required structure

GWP opened the pull request:
*Issue #, if available:*
A CICD deploy system needs to be added to satisfy the Solutions Builder requirements, while maintaining the yarn based dev deploy system that is already in place.

*Description of changes:*

This change is adding the CICD deployment system required by the Solutions Builder team, while preserving the Dev Deploy using Yarn. To test this, please deploy the solution twice, once using yarn, and once using the cloudformation CICD template. The structure required by the Solutions site is shown here: http://contribute.awssolutionsbuilder.com/getting-started.html


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

alexchirayath commented on README.md:
```diff
@@ -1,39 +1,108 @@
+### Notes
+
+- Do NOT change the `cicd` in package.json. This field is for the deployment system to use in CodePipeline
+- Dude to limitations of CodeCommit, you cannot use this deploy approach if you add a file to the solution that is above 6MB (for good measure, stay below 5MB)
```
NIT : Due

alexchirayath commented on deployment/build-and-deploy-project.sh:
```diff
@@ -0,0 +1,103 @@
+#!/bin/bash
```
Do we need this file or should we add it to .gitignore?

alexchirayath commented on deployment/build-dist/document-understanding-solution.template:
```diff
@@ -0,0 +1,791 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Do we need this file or should we add it to .gitignore?

GWP commented:
> I havent ran the demo yet but have comments for now
> 
> 1. Some changes to the Readme commented
> 2. Do we need the .template and cicd.zip file or should we add it to .gitgnire?
> 3. The appraoch to add CICD was to make it easier (minimal terminal use) for the customer. Seems that we have new additional steps for allowing both deploys to work. Is this also the recommended approach from Solution Builders?
> 4. If by any chance others approve and you merge in your changes, please make sure to rebase on master before pushing your code.
> 
> Will add in more comments once I have bandwidth to run the demo

I've fixed the type, added files to the gitignore, and removed the old build file. Unfortunately, the difficulty with deploying this from the CLI is necessary, as it will allow the SB team to create a one-click deploy in their pipeline. Refer to this published solution for more information:

https://github.com/awslabs/smart-product-solution

alexchirayath commented on source/package.json-e:
```diff
@@ -14,26 +14,31 @@
     "Firefox ESR"
```
Do we need this file?

GWP replied:
No, I've removed it

alexchirayath commented on README.md:
````diff
@@ -1,39 +1,108 @@
+- Now build the distributable:
+
+```
+chmod +x ./deployment/build-s3-dist.sh
````
Could you double check this command?
I cannot find the build-s3-dist.sh script

GWP replied:
File has been added to PR

alexchirayath commented on README.md:
```diff
@@ -1,39 +1,108 @@
+
+_Note:_ You will have to create an S3 bucket with the template 'my-bucket-name-<aws_region>'; aws_region is where you are testing the customized solution. Also, the assets in this bucket should be publicly accessible.
+
+For example, you create a bucket called `document-understanding-bucket-us-east-1`,
```
NIT : Seems that now ther Solutions Builder team has moved to store all solutions in one bucket with each solution separated by prefix as shown athttps://github.com/awslabs/smart-product-solution/blame/master/README.md#L22
We could use another example name such as "my-solutions-bucket-us-east-1"

alexchirayath commented on README.md:
```diff
@@ -1,39 +1,108 @@
 ### Development Deploy Workflow and stacknaming
 
-The `package.json` script node `stackname` sets the stackname for the deploy commands. Throughout development it has been imparative to maintain multiple stacks in order to allow client app development and stack architecture development to work without creating breaking changes. When a new stackname is merged into develop it should have the most up to date deployments.
+The `package.json` script node `stackname` sets the stackname for the deploy commands. Throughout development it has been imperative to maintain multiple stacks in order to allow client app development and stack architecture development to work without creating breaking changes. When a new stackname is merged into develop it should have the most up to date deployments.
```
I know this can be out of the scope of this PR, but have there been any thoughts of using Nested stacks?
We have one master stack that can then spin both the DUS and Client stack for Yarn deploy as well and one yarn destroy.

alexchirayath commented on source/package.json-e:
```diff
@@ -14,26 +14,31 @@
-    "destroy": "AWS_REGION=$npm_package_stack_region cdk destroy -v true -a \"node bin/deploy-client-stack.js\" && AWS_REGION=$npm_package_stack_region cdk destroy -v true",
-    "update-client-stack-variables": "bash ./deployment/custom-deployment/bin/stack-variables-to-client.sh",
-    "update-pdf-lambda-code": "bash ./deployment/custom-deployment/bin/update-pdf-lambda-code.sh",
+    "destroy": "STACKNAME=$npm_package_stack_stackname AWS_REGION=$npm_package_stack_region USER_EMAIL=$npm_package_email cdk destroy -v true $npm_package_stack_stacknameStack $npm_package_stack_stacknameClientStack",
```
Any reason why the command has been changed to this?
Would it beneficial to get the client stack name from  "node bin/deploy-client-stack.js" as done in yarn deploy:client?

alexchirayath commented on source/package.json:
```diff
@@ -0,0 +1,115 @@
+    "express": "^4.17.1",
+    "inquirer": "^6.4.1",
+    "lodash": "^4.17.11",
+    "next": "^8.1.0",
```
We might have to upgrade to 9.2.2 since the prior versions depend on and older version of serialize-javascript that has a vulnerability
https://www.npmjs.com/advisories/1426

GWP replied:
we can test out updating this version and make sure it does not break anything. Worth doing separately from this PR

alexchirayath replied:
Pushing this change to Github without the update will probably open another vulnerability ticket similar to this [one](https://github.com/awslabs/amazon-textract-document-understanding-solution/network/alert/yarn.lock/serialize-javascript/closed)

GWP commented:
I have addressed comments by @alexchirayath. I have tried manually updating the Boto dependency, but it says everything is up to date. The destroy command was changed because CDK tried to destroy a stack with "undefined" name if I did not set the process.env.STACKNAME variable. Updated .gitignore

alexchirayath commented:
> I have addressed comments by @alexchirayath. I have tried manually updating the Boto dependency, but it says everything is up to date. The destroy command was changed because CDK tried to destroy a stack with "undefined" name if I did not set the process.env.STACKNAME variable. Updated .gitignore

In your local deploy, can you verify the website functionalities are working as expected?
Running from this branch breaks the comprehend ICD10 API calls which would break the demo.

As for the naming for the client stack, using ClientStack in the end is more of hardocding which I would prefer to avoid. Could we get thestack name from the config since its deployed?
